### ISSUE TITLE
`Error: Unknown value type` on nested create

### ISSUE THREAD

**Issue by Blitz2145** [prisma/prisma#20192]
### Bug description

After upgrading to prisma v5, we are now seeing the following error when executing a nested create query. We have downgraded to 4.16.2 as a workaround. `DEBUG=prisma*` did not yield any more relevant logs for the error. Will attempt to make a small repro.

```
error - Error: Unknown value type
    at pe (/.../node_modules/@prisma/client/runtime/library.js:26:706)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8408)
    at ia (/.../node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/.../node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/.../node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/.../node_modules/@prisma/client/runtime/libra15:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400) {
  clientVersion: '5.0.0',
  ```

### How to reproduce

1. Run nested create query

### Expected behavior

Nested create query succeeds.

### Prisma information

```ts
datasource db {
  provider = "postgres"
  url      = env("TEST_POSTGRES_URI")
}

generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client"
}

model User {
  uuid           String           @id @default(uuid())
  teamUuid       String           @default(uuid())
  userName       String
  UserTemplateV2 UserTemplateV2[]
}

model UserTemplateV2 {
  uuid   String @id @default(uuid())
  userId String
  from   String
  jsonField Json
  user   User   @relation(fields: [userId], references: [uuid])
}
```

```ts
await prisma.user.create({
  select: { uuid: true, teamUuid: true },
  data: {
    userName: name.trim(),
    UserTemplateV2: {
      create: [
        {
          from: `${user.firstName} ${user.lastName}`,
          jsonField: { href: new URL("http://www.example.com") }
        },
        {
          from: `${user.firstName} ${user.lastName}`,
          jsonField: { href: new URL("http://www.example.com") }
        },
      ],
    },
  },
});
```

### Environment & setup

- OS: Alpine Linux
- Database: PostgreSQL 14
- Node.js version: v16.19.0

### Prisma Version

```
5.0.0
```

**Comment by SevInf** [comment]
Hi @Blitz2145. Could you provide the schema you are using? From the snippet you posted, my best guess is:

```prisma
datasource db {
  provider = "postgres"
  url      = env("TEST_POSTGRES_URI")
}

generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client"
}

model User {
  uuid           String           @id @default(uuid())
  teamUuid       String           @default(uuid())
  userName       String
  UserTemplateV2 UserTemplateV2[]
}

model UserTemplateV2 {
  uuid   String @id @default(uuid())
  userId String
  from   String
  user   User   @relation(fields: [userId], references: [uuid])
}
```

However, I can not reproduce the issue with this schema: `create` successfully finishes, user and two templates are created.  
Also, debug logs from 4.16.2, where the query succeeds would also be helpful.

Thank you!

**Comment by Blitz2145** [comment]
> Hi @Blitz2145. Could you provide the schema you are using? From the snippet you posted, my best guess is:
> 
> ```prisma
> datasource db {
>   provider = "postgres"
>   url      = env("TEST_POSTGRES_URI")
> }
> 
> generator client {
>   provider = "prisma-client-js"
>   output   = "../node_modules/.prisma/client"
> }
> 
> model User {
>   uuid           String           @id @default(uuid())
>   teamUuid       String           @default(uuid())
>   userName       String
>   UserTemplateV2 UserTemplateV2[]
> }
> 
> model UserTemplateV2 {
>   uuid   String @id @default(uuid())
>   userId String
>   from   String
>   user   User   @relation(fields: [userId], references: [uuid])
> }
> ```
> 
> However, I can not reproduce the issue with this schema: `create` successfully finishes, user and two templates are created. Also, debug logs from 4.16.2, where the query succeeds would also be helpful.
> 
> Thank you!

Yes, will do, I had to remove too many internal details from the codebase to post the issue. Trying to come up with a schema I can actually share.

**Comment by SevInf** [comment]
@Blitz2145 if you'd prefer to rather share it privately, you can send an email to schemas(at)prisma.io, we'll take a look.

**Comment by Blitz2145** [comment]
Quick update, the problem is in the value being passed to the json field of `UserTemplateV2` from your schema, updated your schema below. I'm bisecting the json value being passed to narrow down the error.

```
model UserTemplateV2 {
  uuid   String @id @default(uuid())
  userId String
  from   String
  data Json
  user   User   @relation(fields: [userId], references: [uuid])
}
```

**Comment by Blitz2145** [comment]
@SevInf I've updated the description with a schema and query that I think should reproduce the issue. We are passing a URL (https://developer.mozilla.org/en-US/docs/Web/API/URL#instance_methods) value into the json field. Its seems like previously that value was being processed correctly since it implements `toJSON` but maybe that is no longer being called in v5.

**Comment by masterbater** [comment]
https://github.com/prisma/prisma/discussions/20186
I also have this issue Unknown value type error but when using findFirst, findMany or  findUnique this only occurs in production though. I have specific query that is causing this but its hard to figure out and Im still testing it it

**Comment by Jolg42** [comment]
We're looking into this, like @Blitz2145, in the previous versions `.toJSON()` was called by `JSON.stringify()` and it worked "unexpectedly" in JS without type-checking.

With type-checking it is not possible to have `jsonField: { href: new URL("http://www.example.com/") }` for example.

There are 2 options at the moment:
- Downgrade to `4.16.2` and wait for updates here, we are working on a fix and we'll share a special version that you will be able to try.
- Change your queries where you have objects, add `.toJSON()`, like `jsonField: { href: new URL("http://www.example.com/").toJSON() }` 

Note that nothing needs to be changed for `Date`, `Decimal`, `BigInt` or `Buffer`.

**Comment by masterbater** [comment]
> We're looking into this, like @Blitz2145, in the previous versions `.toJSON()` was called by `JSON.stringify()` and it worked "unexpectedly" in JS without type-checking.
> 
> With type-checking it is not possible to have `jsonField: { href: new URL("http://www.example.com/") }` for example.
> 
> There are 2 options at the moment:
> 
> * Downgrade to `4.16.2` and wait for updates here, we are working on a fix and we'll share a special version that you will be able to try.
> * Change your queries where you have objects, add `.toJSON()`, like `jsonField: { href: new URL("http://www.example.com/").toJSON() }`
> 
> Note that nothing needs to be changed for `Date`, `Decimal`, `BigInt` or `Buffer`.



> We're looking into this, like @Blitz2145, in the previous versions `.toJSON()` was called by `JSON.stringify()` and it worked "unexpectedly" in JS without type-checking.
> 
> With type-checking it is not possible to have `jsonField: { href: new URL("http://www.example.com/") }` for example.
> 
> There are 2 options at the moment:
> 
> * Downgrade to `4.16.2` and wait for updates here, we are working on a fix and we'll share a special version that you will be able to try.
> * Change your queries where you have objects, add `.toJSON()`, like `jsonField: { href: new URL("http://www.example.com/").toJSON() }`
> 
> Note that nothing needs to be changed for `Date`, `Decimal`, `BigInt` or `Buffer`.

Hopefully a patch would be release soon.

**Comment by masterbater** [comment]
> > We're looking into this, like @Blitz2145, in the previous versions `.toJSON()` was called by `JSON.stringify()` and it worked "unexpectedly" in JS without type-checking.
> > With type-checking it is not possible to have `jsonField: { href: new URL("http://www.example.com/") }` for example.
> > There are 2 options at the moment:
> > 
> > * Downgrade to `4.16.2` and wait for updates here, we are working on a fix and we'll share a special version that you will be able to try.
> > * Change your queries where you have objects, add `.toJSON()`, like `jsonField: { href: new URL("http://www.example.com/").toJSON() }`
> > 
> > Note that nothing needs to be changed for `Date`, `Decimal`, `BigInt` or `Buffer`.
> 
> Yea, we just changed our queries to call .toJSON() on the values and we were able to roll-forward to v5.

Solve this by converting mongodb ObjectId to string. I received ObjectId from change streams then pass it to prisma queries causing this error.

**Comment by janpio** [comment]
@masterbater Can you please give us some more information on what exactly the schema and the Prisma Client query looked like and which value you had to cast? This sounds different than the `URL()` example from @Blitz2145 - and we would really like to understand this. Thanks!

**Comment by masterbater** [comment]
> @masterbater Can you please give us some more information on what exactly the schema and the Prisma Client query looked like and which value you had to cast? This sounds different than the `URL()` example from @Blitz2145 - and we would really like to understand this. Thanks!

This cause error in findUnique, findMany or findFirst when I was passing an ObjectId in id field

```typescript
const deviceId = ObjectId("507f1f77bcf86cd799439011"); //Simulates deviceId in change streams
  const device = await this.prisma.Devices.findFirst({
          where: { id: deviceId.toString() },
          select: { device: true },
        });
```
You wont have any issue with this if you only use prisma all throughout your application

**Comment by SevInf** [comment]
Hi @Blitz2145 and @masterbater.
I know you mentioned that you've used workarounds to update your projects, but in case you'd like to revert that in future update, i'd like to let you know that it's fixed and will be released as a part of Prisma 5.1. If you'd like to test the fix earlier, you can use dev snapshot version `5.1.0-dev.18`. We don't recommend using in production, but it should be good enough to verify that fix works for you.
Thank you!

**Comment by wtachau** [comment]
FYI still seeing 
```sh
Error: Unknown value type
    at pe (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:26:706)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8408)
    at ia (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8400)
    at Ap (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8680)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8063) {
  clientVersion: '5.1.0-integration-fix-client-types-mixed-input-types.1'
}
```
after updating to version `5.1.0-dev.18`

**Comment by SevInf** [comment]
@wtachau client version in your error message is `5.1.0-integration-fix-client-types-mixed-input-types.1`, not `5.1.0-dev.18`. Please, make sure to update both `@prisma/client` and `prisma` packages and re-run `prisma generate` after doing so.

**Comment by wtachau** [comment]
Oh interesting, I put
```json
{
 "prisma": "^5.1.0-dev.18",
 "@prisma/client": "^5.1.0-dev.18",
}
```
in my package.json, re-ran `yarn` and `prisma generate` and that's what got generated. I'll specify `5.1.0-dev.18` exactly later and see if it still happens.

**Comment by wtachau** [comment]
Just to follow up here— I tried again specifying `5.1.0-dev.18` exactly, and now the client version does match. I get a new error (below)— let me know if you'd like me to create an issue for it or if you all are already aware/ it might get fixed by some other in progress tickets.
```
Invalid value for argument `inSafeRange`: We could not serialize [object Function] value. Serialize the object to JSON or implement a ".toJSON()" method on it.
    at Ur (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:5852)
    at ht.throwValidationError (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:9420)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8486)
```

**Comment by SevInf** [comment]
@wtachau this error means that `isSafeRange` parameter is a function and not something that can be passed to the query. Same error should report exactly where in your query it occurs. You should double check what you are passing to the prisma query and ensure it is one of the built-ins supported prisma types: all JS primitives except symbols, `Date`, `Buffer` and `Decimal` objects, field references, or, after the fix, anything with `toJSON` method.

**Comment by wtachau** [comment]
@SevInf Thanks— figured out that I was storing a third-party object in a `Json` field (which is what confused me about `isSafeRange`, since that's not in my code base). The error I got is below. It would be helpful if:
- the stack trace told me where in my code the error was happening, or where in the query (it just tells me the query itself)
- the migration docs mentioned that you have to explicitly call `toJSON`/`JSON.stringify` on objects that you're storing in `Json` fields, since you didn't have to do that before.

```
PrismaClientValidationError:
Invalid `prisma.validation.update()` invocation:

{
  where: {
    id: "53d11bf3-7656-4cee-a97a-e165d1700151"
  },
  data: <omitted>
}

Invalid value for argument `inSafeRange`: We could not serialize [object Function] value. Serialize the object to JSON or implement a ".toJSON()" method on it.
    at Ur (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:5852)
    at ht.throwValidationError (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:9420)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8486)
    at ia (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8936)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8476)
    at ia (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8936)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8476)
    at ia (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8936)
    at na (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:8476)
    at Ap (/Users/williamtachau/src/map/node_modules/@prisma/client/runtime/library.js:115:9054) {
  clientVersion: '5.1.0-dev.18'
}
```

**Comment by SevInf** [comment]
@wtachau what exactly was that third party object? If it has `toJSON` method, explicit serialization should not be necessary.

**Comment by wtachau** [comment]
The object is a result from `https://www.npmjs.com/package/neo4j-driver` , which is just an `"object"`. I suppose somewhere nested in it is an instance of a Class. At any rate, given that this used to work and now doesn't, I would have expected to see it mentioned in the migration guide (maybe it's in there and I missed it).

**Comment by Jolg42** [comment]
@wtachau We did not mention in the migration guide because it's an unexpected bug, which is now fixed and will be included in the next release (5.1.0 scheduled for August 1st).
